File: day03/main.go
// Advent of Code 2023, Day 03
//
// Given a surface of digits and special characters, find any numbers that
// are adjacent to special characters. For Part 2, find any pairs of numbers
// that are adjacent to the same asterisk.
//
// AK, 3 Dec 2023 (60 mins)

package main

import (
	"fmt"
)

// A row & column position (required because of compound map key)
type Position struct {
	r, c int
}

// Global dictionary of part number lists, mapped by the position of the "gear"
// the are adjacent to (required for Part 2)
var Gears map[Position][]int

func main() {

	// Read the input file into a list of strings
	fname := "sample.txt"
	fname = "input.txt"
	lines := readLines(fname)

	// Process each line, by looking for numbers (consecutive digits), and
	// the characters in their bounding box. For Part 1, just see if there
	// are any adjacent special characters. For Part 2, keep track of the
	// numbers that are found adjacent to any asterisks.
	var ans1, ans2 int           // the answers for parts 1 & 2, initialized to zero
	Gears = map[Position][]int{} // initialize global map for Part 2
	for ri, l := range lines {

		// Look for start of the next number. c0 will be the starting position
		// of a number, c1 one past the end of the number.
		var c0, c1 int    // both initialized to zero, thanks Go!
		for c0 < len(l) { // continue to past end of line

			// Skip if not a digit
			if !isDigit(l[c0]) {
				c0++
				continue
			}

			// Ok, start of number found, now find the end of the number
			for c1 = c0; c1 < len(l) && isDigit(l[c1]); c1++ {
			}

			// Extract the number found
			partNo := atoi(string(l[c0:c1]))

			// Part 1: add up part number if it is adjacent to any 'symbols'
			// (this also builds up a map of numbers found adjacent to each
			// asterisk, for Part 2)
			if adjacentToSymbols(lines, ri, c0, c1) {
				ans1 += partNo
			}

			// Move starting position past this number
			c0 = c1
		}

	}
	fmt.Println("Part 1:", ans1) // s/b 560670

	// For Part 2, look at each "gear" that was found, and if there were
	// two part numbers adjacent to it, add their product to the answer
	for _, partNos := range Gears {
		if len(partNos) == 2 {
			ans2 += partNos[0] * partNos[1]
		}
	}
	fmt.Println("Part 2:", ans2) // s/b 91622824
}

// Determine if number is adjacent to any symbols
func adjacentToSymbols(lines []string, ri, c0, c1 int) bool {
	var isAdjacent bool                     // initialized to false
	for r := max(ri-1, 0); r <= ri+1; r++ { // row above and below
		if r >= len(lines) {
			break
		}
		for c := max(c0-1, 0); c <= c1 && c < len(lines[r]); c++ {

			// If character at this position is a symbol, we're true for Part 1
			ch := lines[r][c]
			if isSymbol(ch) {
				isAdjacent = true
			}

			// If the symbol is a star, add this number to the list of
			// numbers adjacent this 'gear' (for Part 2)
			if ch == '*' {
				l := lines[ri]
				partNo := atoi(string(l[c0:c1]))
				pos := Position{r, c}
				Gears[pos] = append(Gears[pos], partNo)
			}
		}
	}
	return isAdjacent // return value is used for Part 1
}

// A character is a symbol if it's not a digit or period
func isSymbol(c byte) bool {
	return c != '.' && !isDigit(c)
}

// Is this character a digit?
func isDigit(c byte) bool {
	return c >= '0' && c <= '9'
}

File: day03/main.rs
// Advent of Code 2023, Day 03
//
// Given a surface of digits and special characters, find any numbers that
// are adjacent to special characters. For Part 2, find any pairs of numbers
// that are adjacent to the same asterisk.
//
// AK, Rust version 3 Oct 2024

use std::fs;
use regex::Regex; // Cargo.toml: regex = "1.3.9"

struct Symbol {
    y: usize,
    x: usize,
    symbol: char,
}

struct SymbolNums {
    s: Symbol,
    nums: Vec<i32>,
}

fn main() {
    
    // Read input file into a list of strings
    //let fname = "sample.txt";
    let fname = "input.txt";
    let data = fs::read_to_string(fname)
        .expect("Unable to read");
    let lines: Vec<_> = data.lines().collect();

    // Regular expression for a sequence of digits
    let re = Regex::new(r"\d+").unwrap();

    // Process each line
    let mut part1 = 0;
    let mut all_symbols = Vec::<SymbolNums>::new();
    for i in  0..lines.len() {

        // Find all sequences of digits in this line
        let l = lines[i];
        for mat in re.find_iter(l) {

            // Get position and value of match
            let p0 = mat.start();
            let p1 = mat.end();
            let num =  &l[p0..p1];
            let n: i32 = num.parse::<i32>().unwrap();

            // Find the symbols adjacent to this number
            let adj_symbols = adj_symbols(i, p0, p1, &lines); 

            // Part 1: add up numbers that are adjacent to any symbol
            if adj_symbols.len() > 0 {
                part1 += n;
            }

            // For part 2, accumulate list of part numbers adjacent to each symbol found
            for s in adj_symbols {

                // Find the symbol in list, add number to list if found
                match find_symbol(&s, &all_symbols) {
                    Some(i) => all_symbols[i].nums.push(n),
                    None => {
                        let mut s1 = SymbolNums{s: s, nums: Vec::<i32>::new()};
                        s1.nums.push(n);
                        all_symbols.push(s1);
                    }
                }
            }
        }
       
    }

    // Answer for Part 1
    println!("Part 1 (s/b 4361, 560670) = {}", part1);

     // Part 2: add up products of pairs of part numbers that are adjacent to the 
     // same asterisk
     let mut part2 = 0;
     for s in all_symbols {
        if s.s.symbol == '*' && s.nums.len() == 2 {
            part2 += s.nums[0] * s.nums[1];
        }
     }
     println!("Part 2 (s/b 467835, 91622824) = {}", part2);
}


// Find symbol in a list, return index to the symbol found
// TODO: tried to return pointer but did not succeed
fn find_symbol(s: &Symbol, ss: &Vec<SymbolNums> ) -> Option<usize> {
    for i in 0..ss.len() {
        let s1 = &ss[i].s;
        if s1.x == s.x && s1.y == s.y && s1.symbol == s.symbol {
            return Some(i);
        }
    }
    None
}

// Check if there are any symbols adjacent to sequence of characters on given line,
// return a list of the symbols found with their coordinates
fn adj_symbols(line_no: usize, start: usize, end: usize, lines: &Vec<&str>) -> Vec<Symbol> {

    // We will return a list of all the symbols this number is adjacent to,
    // with their x,y position
    let mut symbols = Vec::new();

    // Row before: check all positions from before number to one after
    if line_no > 0 {
        let l = lines[line_no - 1];
        let mut x = start;
        if x > 0 {
            x -= 1;
        }
        while x < l.len() && x <= end {
            let c = l.chars().nth(x).unwrap(); // WTF???!!!
            if issymbol(c) {
                let symb = Symbol{y: line_no-1, x: x, symbol: c};
                symbols.push(symb);
            }
            x += 1;
        }
    }

    // Row after: check all positions from before number to one after
    if line_no < lines.len() - 1 {
        let l = lines[line_no + 1];
        let mut x = start;
        if x > 0 {
            x -= 1;
        }
        while x < l.len() && x <= end  {
            let c = l.chars().nth(x).unwrap(); // WTF???!!!
            if issymbol(c) {
                let symb = Symbol{y: line_no+1, x: x, symbol: c};
                symbols.push(symb);
            }
            x += 1;
        }
    }

    // This row: check character before and after
    let l = lines[line_no];
    if start > 0 {
        let c = l.chars().nth(start-1).unwrap();
        if issymbol(c) {
            let symb = Symbol{y: line_no, x: start-1, symbol: c};
            symbols.push(symb);
        }
    }
    if end < l.len() {
        let c = l.chars().nth(end).unwrap();
        if issymbol(c) {
            let symb = Symbol{y: line_no, x: end, symbol: c};
            symbols.push(symb);
        }
    }

    // Return list of symbols adjacent to this number
    symbols
}

// Check if character is a symbol, e.g., not a period or digit
fn issymbol(c: char) -> bool {
    c != '.' && !isdigit(c)
}

// Check if character is a digit
fn isdigit(c: char) -> bool {
    c >= '0' && c <= '9'
}
